src/core/components/pin.js

```diff
@@ -52,30 +53,49 @@ module.exports = (self) => {
   const recursiveKeys = () =>
     Array.from(recursivePins).map(key => new CID(key).buffer)
 
-  function getIndirectKeys (callback) {
-    const indirectKeys = new Set()
-    eachLimit(recursiveKeys(), concurrencyLimit, (multihash, cb) => {
-      dag._getRecursive(multihash, (err, nodes) => {
+  function walkDag ({ cid, preload = false, onCid = () => {} }, cb) {
+    const q = queue(function ({ cid }, done) {
+      dag.get(cid, { preload }, function (err, result) {
         if (err) {
-          return cb(err)
+          return done(err)
         }
 
-        map(nodes, (node, cb) => util.cid(util.serialize(node), {
-          cidVersion: 0
-        }).then(cid => cb(null, cid), cb), (err, cids) => {
-          if (err) {
-            return cb(err)
-          }
+        onCid(cid)
 
-          cids
-            .map(cid => cid.toString())
-            // recursive pins pre-empt indirect pins
-            .filter(key => !recursivePins.has(key))
-            .forEach(key => indirectKeys.add(key))
+        if (result.value.Links) {
+          q.push(result.value.Links.map(link => ({
+            cid: link.Hash
+          })))
+        }
 
-          cb()
-        })
+        done()
       })
+    }, concurrencyLimit)
+    q.drain = () => {
+      cb()
+    }
+    q.error = (err) => {
+      q.kill()
+      cb(err)
+    }
+    q.push({ cid })
+  }
+
+  function getIndirectKeys (callback) {
```

Comment:
You need to pass the `preload` option from `pin.ls` here and then pass it to `walkDag`

Reply:
Done!